src/Http/Middlewares/DeprecatedRoute.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace Jobilla\DeprecatedRoutes\Http\Middlewares;
+
+use Carbon\Carbon;
+use Illuminate\Http\Request;
+use Illuminate\Http\Response;
+use Illuminate\Support\Facades\Config;
+use Illuminate\Support\Facades\Event;
+use Illuminate\Support\Facades\Log;
+use Jobilla\DeprecatedRoutes\Events\RouteIsDeprecated;
+
+class DeprecatedRoute
+{
+    /**
+     * @var string
+     */
+    public const HEADER_NAME = 'X-Deprecated-At';
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure  $next
+     * @param  mixed  $deprecatedAt
+     * @return mixed
+     */
+    public function handle(Request $request, \Closure $next, $deprecatedAt): Response
+    {
+        /** @var \Illuminate\Http\Response $response */
+        $response = $next($request);
+
+        $this->logDeprecatedRequest($request);
+        $this->fireDeprecationEvent($request, $response);
+
+        $response->header(static::HEADER_NAME, $this->getTimestampString($deprecatedAt));
+
+        return $response;
+    }
+
+    protected function logDeprecatedRequest(Request $request)
+    {
+        $level = Config::get('deprecated_routes.log_level');
+
+        if ($level === false) {
+            return;
+        }
+
+        Log::log($level, 'Deprecated route on ' . $request->path() . ' was requested.');
+    }
+
+    protected function fireDeprecationEvent(Request $request, Response $response)
+    {
+        $enabled = Config::get('deprecated_routes.fire_event');
+
+        if ($enabled !== true) {
+            return;
+        }
+
+        Event::dispatch(new RouteIsDeprecated($request, $response));
+    }
+
+    protected function getTimestampString($deprecatedAt): string
+    {
+        try {
+            $datetime = Carbon::createFromFormat('Y-m-d', $deprecatedAt)->startOfDay();
+        } catch (\Exception $e) {
+            throw new \Exception('Deprecate date has to be a date string in Y-m-d format.', 0, $e);
```

Comment:
Should we make it a specific exception rather than throwing a generic?

Reply:
I'm not sure. How would it be beneficial to have a custom defined exception?

Reply:
With custom defined exceptions it allows the end-developer to handle those exceptions themselves, and things like Laravel Ignition would be able to provide their own solutions to the exception based on the exception type itself

Reply:
Any thoughts on the latest changes I just pushed?

Reply:
Looks good on this bit, just 1 comment left.